commitizen/changelog.py

```diff
@@ -131,3 +150,82 @@ def generate_block_tree(block: List[str]) -> Dict:
 def generate_full_tree(blocks: Iterable) -> Iterable[Dict]:
     for block in blocks:
         yield generate_block_tree(block)
+
+
+def get_commit_tag(commit: GitProtocol, tags: List[GitProtocol]) -> Optional[GitTag]:
```

Comment:
@Lee-W maybe you can help me here. I wanted to create a `Protocol` here, but mypy is failing to detect it. 
By Protocol I understood that I could say: "What's being received as GitProtocol should have the attribute `rev`". But for some reason, mypy sees it as a normal type.
Any idea how can I annotate this?

Reply:
We could use `def get_commit_tag(commit: GitProtocol, tags: List[GitProtocol]) -> Optional[GitProtocol]:`. I'm still learning this concept haha 😆 The protocol seems to be a less strict type, so if you annotate it as a stricter type (i.e., `GitTag`), it cannot guarantee that's correct.

BTW, could you rebase master back to the `command-changelog`  branch. I notice that I forget to do so for a while and you just push some commits to the branch.

Reply:
I removed my commits from `command-changelog`, because the idea was to make a PR to that branch haha.  But I'l rebase it and this one then

Reply:
But the thing regarding Protocol, is that is not working as I intend. Maybe I have to configure mypy to be less strict... I haven't check that.

Even the example from the [PEP](https://www.python.org/dev/peps/pep-0544/#type-and-class-objects-vs-protocols) is linting errors with mypy:

```
from typing import Any, Protocol

class ProtoA(Protocol):
    def meth(self, x: int) -> int: ...
class ProtoB(Protocol):
    def meth(self, obj: Any, x: int) -> int: ...

class C:
    def meth(self, x: int) -> int: ...

a: ProtoA = C  # Type check error, signatures don't match!
b: ProtoB = C  # This should be OK but still lints an error

```

I'm also familirizing with it, will try to find a solution

Reply:
I already rebase the master branch back to the `command-changelog`.

The warning does seem to be related to a stricter mypy setting. It still failed when I use a less strict config.

It seems to be an error on the page. By changing the last 2 lines, it will pass the linter. It makes sense because a and b are an instance of `C` instead of the type `C`

```python
a: ProtoA = C()
b: ProtoB = C()
```

Back to the commitizen part, I still don't think mypy can infer `def get_commit_tag(commit: GitProtocol, tags: List[GitProtocol]) -> Optional[GitTag]:`. Because it could be `GitTag`, `GitCommit`, `GitObject`  or even other kinds of object that has the same signature. mypy could not be sure which type it should be.